### PR TITLE
when node reboot, ignoring cmdDel request

### PR DESCRIPTION
**Pull request by wangwangyusur**
when node reboot, ignoring cmdDel request
**- What this PR does and why is it needed**

Created a pod, yaml is as follows, when the node restarted, kubelet will continuously delete the pause container several times, all down the execution, and when the node restarted the pod re-created, the pod's sleep time is up, the kubelet will also delete the pause container, and after that, down the execution until the report of the error.

This modification omits unnecessary deletion operations.
```
apiVersion: v1
kind: Pod
metadata:
  namespace: ovn-poc
  name: wang
  annotations:
    v1.multus-cni.io/default-network: kube-system/test
    ovn.kubernetes.io/logical_switch: ovn-default
spec:
  nodeName: k8snode1
  containers:
  - name: iperf3
    image: docker.io/library/ubuntu1
    imagePullPolicy: IfNotPresent
    command:
    - sleep
    - "1d"
```
The logs are as follows:
when node reboot:
```
I1208 03:14:31.210278    1846 cni.go:296] [ovn-poc/wang1 1f2c424482d339a10f4ad12699163e465edebac0a27c811a8f24d932552b653e] DEL starting CNI request [ovn-poc/wang1 1f2c424482d339a10f4ad12699163e465edebac0a27c811a8f24d932552b653e]

I1208 03:14:31.221273    1846 cni.go:317] [ovn-poc/wang1 1f2c424482d339a10f4ad12699163e465edebac0a27c811a8f24d932552b653e] DEL finished CNI request [ovn-poc/wang1 1f2c424482d339a10f4ad12699163e465edebac0a27c811a8f24d932552b653e], result "", err failed to get container namespace for pod ovn-poc/wang1: failed to Statfs "": no such file or directory
```
The pod sleep time is up and cni reports an error:
```
I1208 03:22:31.446310    1846 cni.go:296] [ovn-poc/wang1 958e8f0fee39c4422048ce4cc225849885fcd60f8dcf09a74cefc8662fcffcce] DEL starting CNI request [ovn-poc/wang1 958e8f0fee39c4422048ce4cc225849885fcd60f8dcf09a74cefc8662fcffcce]

I1208 03:22:31.456791    1846 cni.go:317] [ovn-poc/wang1 958e8f0fee39c4422048ce4cc225849885fcd60f8dcf09a74cefc8662fcffcce] DEL finished CNI request [ovn-poc/wang1 958e8f0fee39c4422048ce4cc225849885fcd60f8dcf09a74cefc8662fcffcce], result "", err failed to get container namespace for pod ovn-poc/wang1: failed to Statfs "": no such file or directory
```
Environmental information:
```
kubelet --version
Kubernetes v1.24.2

containerd --version
containerd github.com/containerd/containerd v1.6.19 1e1ea6e986c6c86565bc33d52e34b81b3e2bc71f

kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.2", GitCommit:"f66044f4361b9f1f96f0053dd46cb7dce5e990a8", GitTreeState:"clean", BuildDate:"2022-06-15T14:20:54Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
```

**- Special notes for reviewers**

When calling cmdDel, determine if podRequest.NetNs is empty, and Ignoring this request.

**- How to verify it**

Creating a pod, reboot the work node, when pod sleeping time has come, The question appeared.

**- Description for the changelog**

when node reboot, ignoring cmdDel request

### CODE REVIEW

**Comment by tssurya** [comment]
@wangwangyusur thanks for the PR, Curiosity q :Why is the test being done on 1.24 Kube and not a later release? Does this happen in latest versions as well?

@zshi-redhat : does multus do anything similar where it ignores del's with empty netns?

**Comment by zshi-redhat** [comment]
> @wangwangyusur thanks for the PR, Curiosity q :Why is the test being done on 1.24 Kube and not a later release? Does this happen in latest versions as well?
> 
> @zshi-redhat : does multus do anything similar where it ignores del's with empty netns?

@tssurya multus continues to process the cmdDel in case netns doesn't exist, relevant PRs:
https://github.com/k8snetworkplumbingwg/multus-cni/pull/326
https://github.com/k8snetworkplumbingwg/multus-cni/pull/269
This behavior is implied in CNI spec: https://github.com/containernetworking/cni/blob/main/SPEC.md#del-remove-container-from-network-or-un-apply-modifications

File: go-controller/pkg/cni/cni.go [hunk 1]
```diff
@@ -170,6 +170,11 @@ func (pr *PodRequest) cmdDel(clientset *ClientSet) (*Response, error) {
 		return nil, fmt.Errorf("required CNI variable missing")
 	}
 
+	if pr.Netns == "" {
+		klog.Infof("The cmdDel request for pod %s/%s Netns is nil. Ignoring this request.", namespace, podName)
+		return response, nil
```
Comment (tssurya) [inline go-controller/pkg/cni/cni.go:175]: `Plugins should generally complete a DEL action without error even if some resources are missing. For example, an IPAM plugin should generally release an IP allocation and return success even if the container network namespace no longer exists, unless that network namespace is critical for IPAM management.`

https://github.com/containernetworking/cni/blob/main/SPEC.md#del-remove-container-from-network-or-un-apply-modifications

If we return here we won't be processing anything after this, is that the desired behaviour? PR title says "ignore" but really we are killing the delete totally and not continuing to see if something else needs cleanup..

**Review by tssurya — COMMENTED** [review]
thanks @zshi-redhat for the great links! that helps.!

File: go-controller/pkg/cni/cni.go (hunk 1 above)
Comment (wangwangyusur) [reply go-controller/pkg/cni/cni.go:175]: Thanks for your reply.
When dpuhost side reboot and the pod's sleep time is expired，the cmdDel function is called. And at this point the kernel Netns have been cleared after reeboot.
The cmdDel function does only two main things:  delete the DPU connection-details annotation and UnconfigureInterface.
In our case, since only the dpu host is restarted, there is no change on the dpu Soc side, the DPU connection-details annotation should not be cleared.
In UnconfigureInterface function，if Netns is nil，returns "failed to get container namespace".
In conclusion, when Netns is not present, the latter process should not be executed.

Comment (wangwangyusur) [reply go-controller/pkg/cni/cni.go:175]: Thanks for your reply.
When dpuhost side reboot and the pod's sleep time is expired，the cmdDel function is called. And at this point the kernel Netns have been cleared after reeboot.
The cmdDel function does only two main things:  delete the DPU connection-details annotation and UnconfigureInterface.
In our case, since only the dpu host is restarted, there is no change on the dpu Soc side, the DPU connection-details annotation should not be cleared.
In UnconfigureInterface function，if Netns is nil，returns "failed to get container namespace".
In conclusion, when Netns is not present, the subsequent operations should not be performed.

**Comment by wangwangyusur** [comment]
> @wangwangyusur thanks for the PR, Curiosity q :Why is the test being done on 1.24 Kube and not a later release? Does this happen in latest versions as well?
> 
> @zshi-redhat : does multus do anything similar where it ignores del's with empty netns?

We tried 1.27.1 version，having the same problem.
Below are the log messages.

[ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e] DEL starting CNI request [ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e]

Tear down interface ({Command:DEL PodNamespace:ovn-poc PodName:wang1 PodUID:9d8c5a9c-7e4a-485d-827e-25344c6f1bfa SandboxID:1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e Netns: IfName:eth0 CNIConf:0xc00097c100 timestamp:{wall:13932765867401127166 ext:534186348788 loc:0x2e1dd00} ctx:0xc00032df80 cancel:0x49a760}) for pod ovn-poc/wang1

[ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e] DEL finished CNI request [ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e], result "", err failed to get container namespace for pod ovn-poc/wang1: failed to Statfs "": no such file or directory

￼

version info:

root@node1:~# kubelet --version
Kubernetes v1.27.1
root@node1:~#
root@node1:~# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-14T13:20:04Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by wangwangyusur** [comment]
> @wangwangyusur thanks for the PR, Curiosity q :Why is the test being done on 1.24 Kube and not a later release? Does this happen in latest versions as well?
> 
> @zshi-redhat : does multus do anything similar where it ignores del's with empty netns?

We tried 1.27.1 version，having the same problem.
Below are the log messages.

[ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e] DEL starting CNI request [ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e]

Tear down interface ({Command:DEL PodNamespace:ovn-poc PodName:wang1 PodUID:9d8c5a9c-7e4a-485d-827e-25344c6f1bfa SandboxID:1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e Netns: IfName:eth0 CNIConf:0xc00097c100 timestamp:{wall:13932765867401127166 ext:534186348788 loc:0x2e1dd00} ctx:0xc00032df80 cancel:0x49a760}) for pod ovn-poc/wang1

[ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e] DEL finished CNI request [ovn-poc/wang1 1987a9cd9f52d485b940de685c685f0160c4ce04ff95d1bc9af8e28ade2ff88e], result "", err failed to get container namespace for pod ovn-poc/wang1: failed to Statfs "": no such file or directory

￼

version info:

root@node1:# kubelet --version
Kubernetes v1.27.1

root@node1:~# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-14T13:20:04Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by wangwangyusur** [comment]
@tssurya @trozet @dcbw @girishmg @jcaamano 
Please take a look. Thank you.

**Comment by npinaeva** [comment]
I believe this is stale
